doc/reference/src/low-level-execution-model.pdc

```diff
@@ -111,276 +111,542 @@ The oracle-free abstract algorithm for optimal reduction operates on four node t
+  \\ \\
+  The first being the variable name from a specific lambda.
+  \\ \\
+  The second map is a map from the fan in number to the current status of our traversal of said fan-in.
+  \\ \\
+  We can now a proof by induction, specifically, we shall consider three cases:
+
+  1. $\lambda$ /\ $\mu$
+
+  2. $FanIn$
+
+  3. The rest.
+
+  We will also consider the inductive case first for the following proof.
+  \\ \\
+  Let $rec$ be this recursive algorithm.
+  \begin{tcolorbox}[title= Case 1 - rest, colframe=inner-box-1]
+    The logic for all other cases is simple. Consider some node $P$ with ports $1 \cdots n$.
+    \\ \\
+    Let port n be the parent ADT port.
+    \\ \\
+    By induction we run this algorithm and ports $1 \cdots n-1$ are $α$ equivalent to normal evaluation.
+    \\ \\
+    Thus we can construct ADT $P$ by $P(rec(n_1) \cdots rec(n_{n-1}))$.
+    \\ \\
+    $\therefore$ $P$ is well formed and we get an $α$ equivalent answer.
+  \end{tcolorbox}
+
+  \begin{tcolorbox}[title= Case 2 - $\lambda$ /\ $\mu$, colframe=inner-box-1]
+    The $\lambda$ /\ $\mu$ case is special in that in a Valid EAL-Net the node will be traversed via the primary port via the parent AST and then by the second auxiliary port.
+    \\ \\
+    The second auxiliary port refers to the variable binding.
+    \\ \\
+    When we traverse the $\lambda$ node by the primary port first (this must be the case), we add a new var into the first map.
+    \\ \\
+    Then we run the recursive algorithm on Auxiliary 1, this subterm will only access the second Auxiliary port of this $\lambda$ node.
+    \\ \\
+    If the term is unused, then we will connect an eraser node to the second Auxiliary port.
+    \\ \\
+    This thus preserves the number of free ports.
+    \\ \\
+    By induction the subterms on Auxiliary one are also well formed
+    \\ \\
+    $\therefore$ the Lambda term is $α$ equivalent to the normal evaluation answer.
+  \end{tcolorbox}
+
+  \begin{tcolorbox}[title= Case 3 - FanIn, colframe=inner-box-1]
+    FanIn's are another interesting case, for this case we allocate a map from each FanIn to the traversal status.
+    \\ \\
+    For this, we keep a list of the following types
+    \begin{verbatim}
+     type Status = In FanPort
+                 | Complete FanPort
+     type FanPort = Circle | Star
+    \end{verbatim}
+    Where circle is the first auxiliary port and star is the second.
+
+    \begin{figure}[H]
+        \begin{tikzpicture}
+          \inetcell[]{a}{$f_i$}[5]
+          \draw (-0.29, 0.23) node {$\scalebox{1.9}{$\circ$}$};
+          \draw ( 0.29, 0.29) node {$\scalebox{1.7}{$\star$}$};
+          \cutWire{a.out}{b.north}{}{}
+        \end{tikzpicture}
+    \end{figure}
+
+
+    For this analysis we must consider what port we can enter from, however first we should note that before entering the primary port of a FanIn, that we must have first traversed through either $\star$ or $\circ$ first.
```

Comment:
This sounds like you are keeping a first-in-first-out stack of ⋆or∘ indicators while traversing, is that right?

Reply:
first-in-last out stack, correct.

Reply:
can we say that explicitly?

Reply:
Done, it's above where the type is defined.